Sources/SwiftDiagnostics/DiagnosticsFormatter.swift

```diff
@@ -12,311 +12,811 @@
+  ///
+  /// - `.full`: A single range spanning all the lines will be returned, effectively printing the entire source code alongside diagnostics.
+  ///
+  /// - Parameters:
+  ///   - annotatedSourceLines: An array of `AnnotatedSourceLine`, each of which represents a line
+  ///     of source code and any annotations associated with it.
+  ///
+  /// - Returns: An array of `Range<Int>`, each specifying a range of line numbers that should be
+  ///     printed due to the presence of annotations or the context setting.
+  private func computeLinesToPrintRanges(from annotatedSourceLines: [AnnotatedSourceLine]) -> [Range<Int>] {
+    switch contextRange {
+    case .limited(let contextSizeLimit):
+      return
+        annotatedSourceLines
+        .compactMap { annotatedLine -> Range<Int>? in
+          let lineNumber = annotatedLine.sourceLineNumber
+          if annotatedLine.isFreeOfAnnotations {
+            return nil
+          } else {
+            return Range<Int>(uncheckedBounds: (lower: lineNumber - contextSizeLimit, upper: lineNumber + contextSizeLimit + 1))
+          }
         }
+    case .full:
+      return [Range<Int>(uncheckedBounds: (lower: 0, upper: annotatedSourceLines.count))]
+    }
+  }
 
-        return nil
-      }.joined()
+  /// Appends a decorated line of source code to an annotated source string.
+  ///
+  /// - Parameters:
+  ///   - annotatedSource: The string to which the decorated line will be appended.
+  ///   - annotatedLine: The source line to be decorated.
+  ///   - tree: The syntax tree containing the source line.
+  ///   - sourceLocationConverter: The converter used to find source locations.
+  ///   - indentString: The string used for indentation.
+  ///   - maxNumberOfDigits: The maximum number of digits in line numbers, used for alignment.
+  ///
+  /// This function performs the following tasks:
+  /// 1. Decorates the provided source line with highlights.
+  /// 2. Creates a right-aligned line number prefix for improved readability.
+  /// 3. Appends the decorated source line to the `annotatedSource` string.
+  /// 4. If an additional highlighted line exists, it is also appended to `annotatedSource`.
+  ///
+  /// For example, given:
+  /// - `annotatedLine` contains diagnostic with `foo` highlighted,
+  /// - `indentString` is an empty string,
+  /// - `maxNumberOfDigits` is 3,,
+  /// - `annotatedSource` initially contains an empty string
+  ///
+  /// The `annotatedSource` will be updated to:
+  /// ```
+  ///  12 │ func foo(with bar: String) -> Int {
+  ///     │      ~~~
+  /// ```
+  private func appendDecoratedSourceLine(
+    to annotatedSource: inout String,
+    from annotatedLine: AnnotatedSourceLine,
+    within tree: SyntaxProtocol,
+    via sourceLocationConverter: SourceLocationConverter,
+    indentingWith indentString: String,
+    upToMaxDigits maxNumberOfDigits: Int
+  ) {
+    // 1. Decorate source line with highlights
```

Comment:
Nit: I wouldn’t do `1.` etc. Just makes it harder to change if we ever need to introduce anything in between 1. and 2.

Reply:
Removed.